src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c

```diff
@@ -1615,6 +1556,152 @@ static long icap_download_clear_bitstream(struct icap *icap)
+	void *header;
+	XHwIcap_Bit_Header bit_header = { 0 };
+	struct mailbox_req mbreq = { MAILBOX_REQ_CHG_SHELL, };
+	int ret;
+
+	mutex_lock(&icap->icap_lock);
+	if (icap->rp_bit) {
+		ICAP_ERR(icap, "previous dowload in progress");
+		mutex_unlock(&icap->icap_lock);
+		return -EBUSY;
+	}
+
+	section = get_axlf_section_hdr(icap, axlf, DTC);
+	if (!section) {
+		ICAP_ERR(icap, "did not find DTC section");
+		ret = -EINVAL;
+		goto failed;
+	}
+
+	header = (char *)axlf + section->m_sectionOffset;
+
+	if (fdt_check_header(header) || fdt_totalsize(header) >
+			section->m_sectionSize) {
+		ICAP_ERR(icap, "Invalid DTC");
+		ret = -EINVAL;
+		goto failed;
+	}
+
+	icap->rp_fdt = vmalloc(fdt_totalsize(header));
+	if (!icap->rp_fdt) {
+		ICAP_ERR(icap, "Not enough memory for DTC");
+		ret = -ENOMEM;
+		goto failed;
+	}
+	memcpy(icap->rp_fdt, header, fdt_totalsize(header));
+
+	section = get_axlf_section_hdr(icap, axlf, BITSTREAM);
+	if (!section) {
+		ICAP_ERR(icap, "did not find BITSTREAM section");
+		ret = -EINVAL;
+		goto failed;
+	}
+
+	if (section->m_sectionSize < DMA_HWICAP_BITFILE_BUFFER_SIZE) {
+		ICAP_ERR(icap, "bitstream is too small");
+		ret = -EINVAL;
+		goto failed;
+	}
+
+	header = (char *)axlf + section->m_sectionOffset;
+
+	if (bitstream_parse_header(icap, header,
+		DMA_HWICAP_BITFILE_BUFFER_SIZE, &bit_header)) {
+		ICAP_ERR(icap, "parse header failed");
+		goto failed;
+	}
+
+	icap->rp_bit_len = bit_header.HeaderLength + bit_header.BitstreamLength;
+	if (icap->rp_bit_len > section->m_sectionOffset) {
+		ICAP_ERR(icap, "bitstream is too bit");
+		goto failed;
+	}
+
+	icap->rp_bit = vmalloc(icap->rp_bit_len);
+	if (!icap->rp_bit) {
+		ICAP_ERR(icap, "Not enough memory for BITSTREAM");
+		ret = -ENOMEM;
+		goto failed;
+	}
+	memcpy(icap->rp_bit, header, icap->rp_bit_len);
+
+	(void) xocl_peer_notify(xocl_get_xdev(icap->icap_pdev), &mbreq,
+			sizeof(struct mailbox_req));
+	ICAP_INFO(icap, "Notified userpf to program rp");
+
+	mutex_unlock(&icap->icap_lock);
+	return 0;
+
+failed:
+	if (icap->rp_bit) {
+		vfree(icap->rp_bit);
+		icap->rp_bit = NULL;
+		icap->rp_bit_len = 0;
+	}
+	if (icap->rp_fdt) {
+		vfree(icap->rp_fdt);
+		icap->rp_fdt = NULL;
+	}
+
+	mutex_unlock(&icap->icap_lock);
+
+	return ret;
+}
+
+
+static int icap_download_rp(struct platform_device *pdev, int level)
+{
+	struct icap *icap = platform_get_drvdata(pdev);
+	xdev_handle_t xdev = xocl_get_xdev(pdev);
+	int ret = 0;
+
+	mutex_lock(&icap->icap_lock);
+	if (!icap->rp_bit || !icap->rp_fdt) {
+		xocl_xdev_err(xdev, "Invalid reprogram request");
+		ret = -EINVAL;
+		goto failed;
+	}
+
+	ret = xocl_fdt_blob_input(xdev, icap->rp_fdt);
+	if (ret) {
+		xocl_xdev_err(xdev, "failed to parse fdt %d", ret);
+		goto failed;
+	}
+
+	ret = xocl_axigate_freeze(xdev, XOCL_SUBDEV_LEVEL_BLD);
+	if (ret) {
+		xocl_xdev_err(xdev, "freeze blp gate failed %d", ret);
+		goto failed;
+	}
+
+#if 0
+	ret = icap_download(icap, icap->rp_bit, icap->rp_bit_len);
+	if (ret)
+		goto failed;
```

Comment:
should at least try to free axi gate if failed here...

Reply:
True. Will fix this